aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -111,6 +114,19 @@ public int read(byte[] b, int off, int len) throws IOException {
     return bytesRead;
   }
 
+  @Override
+  public void readFully(long position, byte[] buffer, int offset, int length) {
+    GetObjectRequest.Builder requestBuilder = GetObjectRequest.builder()
+        .bucket(location.bucket())
+        .key(location.key())
+        .range(String.format("bytes=%s-%s", position, position + length));
+
+    S3RequestUtil.configureEncryption(awsProperties, requestBuilder);
+
+    ResponseBytes<GetObjectResponse> response = s3.getObject(requestBuilder.build(), ResponseTransformer.toBytes());
+    System.arraycopy(response.asByteArrayUnsafe(), 0, buffer, offset, length);
```

Comment:
Should we be validating that `buffer` is big enough to hold the data from `offset` to `length`?

Reply:
`System.arraycopy` is going to be making that same check, so I feel like all we save is the request for an erroneous case.  Feels like unnecessary validation to me, but if others feel strongly, I can add a `Preconditions::checkPositionIndexes`.

Reply:
I would probably do this because we can catch the problem before making an S3 request. It's not a big deal, but if we can avoid an extra read operation that seems worth a simple check.

Reply:
I added checks since we're now not using the System.arraycopy, but I did take a look at many other implementations of readFully (hadoop, trino, commons) and none of them perform these checks, so we're being extra cautious here.